Sources/NIOCore/NIOTimer.swift

```diff
@@ -0,0 +1,94 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2024 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A type that handles timer callbacks scheduled with ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public protocol NIOTimerHandler {
+    func timerFired(loop: any EventLoop)
+}
+
+/// An opaque handle that can be used to cancel a timer.
+///
+/// Users cannot create an instance of this type; it is returned by ``EventLoop/setTimer(for:_:)-5e37g``.
+///
+/// - Seealso: ``EventLoop/setTimer(for:_:)-5e37g``.
+public struct NIOTimer {
+    @usableFromInline
+    enum Backing {
+        /// A task created using `EventLoop.scheduleTask(deadline:_:)`, used by default for `EventLoop` implementations.
+        case scheduledTask(Scheduled<Void>)
+        /// An identifier for a timer, used by `EventLoop` implementations that conform to `CustomTimerImplementation`.
+        case custom(eventLoop: any NIOCustomTimerImplementation, id: UInt64)
```

Comment:
Instead of storing the event loop here I am wondering if we can just store a `cancel` closure. I think I tried that with `ScheduledTask` once but I am unsure if I got it down to 0 allocations. That might allow us to avoid the whole `UInt64` completely

Reply:
I will try this again, but my intuition is that this will not be as performant. Is avoiding the `UInt64` a goal?

Reply:
I tried this and it resulted in an allocation. I have kept the commits in the history so you can see the attempt and the revert:


- b3e49036c: `feedback(attempted): Store a closure instead of UInt64`
- fb5e83536: `feedback(reverted, allocates): Store a closure instead of UInt64`